### ISSUE TITLE
The learning rate of analog tile for TransferCompound

### ISSUE THREAD

**Issue by chaeunl** [IBM/aihwkit#131]
## Description and motivation

I am trying to change the learning rate of an analog tile using `nn.AnalogLinear.analog_tile.set_learning_rate()`. However, for the case of `TransferCompound`, I think there is no way to change the learning rate for each devices defined in `TranferCompound`. For instance, if I define a `TransferCompound` having two devices, A and C (supposing that each device is defined in different tiles), then there are no way to adjust learning rate of each tile of A or C, respectively. 

+ questions. I confirmed that the learning rate of an analog tile equals to `dw_min` of the device. I want to know that is there any specific reason to set the learning rate to be equal to `dw_min`?  

## Proposed solution

It would be great for us if a tile is defined with `TransferCompound` device, it has multiple learning rates for each device or the device is defined with multiple tiles.

**Comment by maljoras** [comment]
Hi @chaeunl ,
One can actually have a different LR for the transfer in the compound. The first learning rate (on A) is defined by the learning rate of the overall device (the SGD learning rate). The second is given by the argument `transfer_lr` which can be chosen to be scaled with the SGD lr by setting `scale_transfer_lr=True`. If you have more than 2 devices in the compound, there is also the option to set all LRs individually by using `transfer_lr_vec` which is then a list of the learning rates (the first for the first transfer, the second for the second and so on).  See here [here](https://aihwkit.readthedocs.io/en/latest/api/aihwkit.simulator.configs.devices.html?highlight=TransferCompound#aihwkit.simulator.configs.devices.TransferCompound) for a more detailed description of all options. 

See also the example [in the documentation](https://aihwkit.readthedocs.io/en/latest/using_simulator.html?highlight=TransferCompound#transfer-compound-device) where a number of these parameters are set in the configuration.

The learning rate of an analog tile is no necessarily `dw_min`. However, `dw_min` is related to the minimal amount of change the device can have, so a value smaller that can only occur on average by the stochastic pulsing probability is below 1. Also larger updates can happen in one update process, by sending more than 1 pulse. The number of pulses is automatically scaled by the requested learning rate in case `update_bl_management=True` and `update_management=True`. The maximal number of pulses per update can be set with the `desired_bl` property.  See [here](https://aihwkit.readthedocs.io/en/latest/api/aihwkit.simulator.configs.utils.html?highlight=updateparameter#aihwkit.simulator.configs.utils.UpdateParameters.update_bl_management)

**Comment by chaeunl** [comment]
@maljoras , Thank you for your quick reply.
Yes, I understood what you are talking about. What I am trying is to make all different `LinearStepDevice` which composes a `TransferCompound` have same allowed range regardless of its slope. The figure shows what I am trying to at below: 
![image](https://user-images.githubusercontent.com/29699502/107989851-789bcd00-7016-11eb-9199-f26a1d12d7cb.png)
To make it possible, I change the two `dw_min` parameters which are  defined in device level (getting parameters from hidden parameters) and tile level (by rpu_config.device.dw_min). I confirmed that if I do not make changes either of them by scaling `dw_min` (multiply it by "slope/base_slope"; for instance, "slope/0.00166", where "slope" is the slope of each device and base_slope is a slope of a baseline device), I cannot get the behavior. For instance, if I only change `dwmin_up` or `dwmin_down` in the hidden parameters, then I got the following result:
![image](https://user-images.githubusercontent.com/29699502/107990248-5d7d8d00-7017-11eb-8ba5-682ef6fbddb7.png)
Or else, if I only change `rpu_config.device.dw_min`, then I got the following result:
![image](https://user-images.githubusercontent.com/29699502/107990438-d5e44e00-7017-11eb-95e4-0d9ea1f59033.png)

So, I have to change both of them. I think changing the `transfer_lr` is another issue in this case because changing it does not guarantee this issue. But, with `TransferCompound`, I cannot modulate `rpu_config.device.dw_min` separately.

If there is misleading of my side, please let me know. Even now, I am getting used to the simulator, but still confused in some points. And so, your comments are very helpful!

**Comment by maljoras** [comment]
I see, that seems to be a quite different issue. Generally, the A device can have different `dw_min` from the C device, however, if you have more than 2 devices in the `TransferCompound` then you cannot set this from the outside currently. In principle it is not a problem to have different meta-parameter (like `dw_min`) for the same devices type,  it is rather to avoid unintended usage eg that one defines different device types that would be incompatible with the pulse updater. So it is turned off by default. However, for two devices, A, C can be different. 

Nevertheless,  as you apparently already have done, the `dw_min_up` and `dw_min_down` can be accessed with the hidden_parameter property. If you do that the `dw_min` needed for the scaling of the update management will then actually be the mean of the `hidden_parameter` instead of the `dw_min` set on the device level. So it should actually take the the correct one.  

The form of defining the device you seek (same `w_max` but different slope)  might be just the `SoftBoundsDevice`. If use:
```
SoftBoundsDevice(dw_min=0.05, dw_min_dtod=2.0, w_min=-1.0, w_max=1.0, w_max_dtod=0.0, w_min_dtod=0.0)
```
(just using the `dw_min_dtod` to get some diversity in the `dw_min`)
then I get something like the following without even touching the hidden parameters. Is that what you are looking for? 
![image](https://user-images.githubusercontent.com/17587387/107993399-59feed00-6fa8-11eb-91aa-9a948fa0d3d6.png)

**Comment by maljoras** [comment]
Indeed if you have a very complicated scenario and assume separated physical tiles anyway, it might be more flexible if you would make your own analog tile (by overloading the AnalogTile) and overloading the forward and update functions in python instead of using the transfer compound. Then you can do whatever you want, for instance having 3 completely different tiles that transfer back and forth and so on, by using the pyTorch functionality.

**Comment by maljoras** [comment]
Hi @chaeunl ,
Indeed, you can also use `LinearStepDevice` for this as `SoftBoundsDevice` is essentially just a specific parameter setting of the `LinearStepDevice`: It sets the parameter to 

```python
LinearStepDevice(
    gamma_up=1.0,
    gamma_down=1.0,
    gamma_up_dtod=0.0,
    gamma_down_dtod=0.05,
    allow_increasing=False,
    mean_bound_reference=False
)
```
This should be the same as the `SoftBoundsDevice()`.  Especially, the `mean_bound_reference=False` is important, otherwise the slope is calculated based on the mean value of the bounds and not the actual bound.

**Comment by maljoras** [comment]
Of course, we would be happy to include your device model into the simulator!
